gtda/curves/features.py

```diff
@@ -0,0 +1,217 @@
+"""Feature extraction from curves."""
+# License: GNU AGPLv3
+
+from copy import deepcopy
+from types import FunctionType
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_is_fitted, check_array
+
+from ._functions import _AVAILABLE_FUNCTIONS, _implemented_function_recipes, \
+    _parallel_featurization
+from ..utils._docs import adapt_fit_transform_docs
+from ..utils.validation import validate_params
+
+
+@adapt_fit_transform_docs
+class StandardFeatures(BaseEstimator, TransformerMixin):
+    """Standard features from multi-channel curves.
+
+    Applies functions to extract features from each channel in each
+    multi-channel curve in a collection.
+
+    Parameters
+    ----------
+    function : string or callable, optional, default: ``max``
+        Function to transform a single-channel curve into scalar features per
+        channel. Implemented functions are [``"identity"``, ``"argmin"``,
+        `"argmax"``, ``"min"``, ``"max"``, ``"mean"``, ``"std"``, ``"median"``,
+        ``"average"``].
+
+    function_params : dict, optional, default: ``None``
+        Additional keyword arguments for `function`. Passing ``None`` is
+        equivalent to passing no arguments. Additionally:
+
+        - If ``function == "average"``, the only argument is `weights`
+          (np.ndarray or None, default: ``None``).
+        - Otherwise, there are no arguments.
+
+    n_jobs : int or None, optional, default: ``None``
+        The number of jobs to use for the computation. ``None`` means 1 unless
+        in a :obj:`joblib.parallel_backend` context. ``-1`` means using all
+        processors.
+
+    Attributes
+    ----------
+    n_channels_ : int
+
+    effective_function_params_ : dict
+        Dictionary containing all information present in `function_params` as
+        well as on any relevant quantities computed in :meth:`fit`.
+
+    """
+    _hyperparameters = {
+        "function": {"type": (str, FunctionType, list, tuple),
+                     "in": tuple(_AVAILABLE_FUNCTIONS.keys()),
+                     "of": {"type": (str, FunctionType, type(None)),
+                            "in": tuple(_AVAILABLE_FUNCTIONS.keys())}},
+        "function_params": {"type": (dict, type(None), list, tuple)},
```

Comment:
This is as in your original design. I think it's ok to have it as it is maybe future-proof?

Reply:
Ah no wait, there's a better reason. It must be None otherwise the default would break either the single function case or the list of functions case!

Reply:
Ah, my bad then! :)
But that's because we treat it that way, right. We could actually allow empty dictionnaries to act as if they were None.
The problem is that right now, the `effective_function_params_` attribute that arises from this treatment, does never contains information it seems. Is that right?

I can also take care of it if you prefer.

Reply:
> We could actually allow empty dictionnaries to act as if they were None.

In principle yes, but I feel it is slightly clearer to have the `None` to avoid confusions.

> The problem is that right now, the effective_function_params_ attribute that arises from this treatment, does never contains information it seems. Is that right?

Wasn't this also the case before? What has changed in your opinion?

Reply:
As I said it' my bad. It was there before. :)

I was also thinking, that `effective_function_` should be an attribute too! As the list of the actual function applied to each channel. In that sense, maybe `effective_function_params_` would make sense to have as an attribute as a list of dictionarry.

Reply:
Yeah so I was very torn about `effective_function_` as I was wondering whether it should always be a list of functions (or `None`) even when `function` was a single function.  I think it is less confusing if `effective_function_` and `effective_function_params_` have always the same type (tuple).  Do you agree that I can:
- add `effective_function_` which is always a tuple of functions or `None` even if the user passes one of the strings or a single function;
- ensure `effective_function_params_` is also always a tuple of (possibly empty) dictionaries.

Reply:
I agree! Why not list btw, given than we also agreed on reducing the complexity of the parallel function with list indexing?

Reply:
I try to be consistent throughout the library with list-like attributes, such as `homology_dimensions_`, to make them all tuples. I think immutable objects should be avoided when they are not meant to be manipulated but just inspected. And I think this is the case of these class attributes, no?

Reply:
Ok, I don't think that's the case all over the library, but why not!

Reply:
Oh ok! I once went through to check for this, so if you find some inconsistencies please let me know ;)

Reply:
I trust you xD